lib/src/migrators/module.dart

```diff
@@ -217,52 +217,91 @@ class _ModuleMigrationVisitor extends MigrationVisitor {
     references.globalDeclarations.forEach(_renameDeclaration);
     var migrated = super.run(stylesheet, importer);
 
-    if (prefixToRemove != null && _needsImportOnly) {
+    if (forwards.contains(ForwardType.importOnly) ||
+        (prefixToRemove != null && _needsImportOnly)) {
       var url = stylesheet.span.sourceUrl;
       var importOnlyUrl = getImportOnlyUrl(url);
-      var tuple = _absoluteUrlToDependency(url, relativeTo: importOnlyUrl);
-      var results = _generateImportOnly(url, tuple.item1);
+      var results = _generateImportOnly(url, importOnlyUrl);
       if (results != null) migrated[importOnlyUrl] = results;
     }
     return migrated;
   }
 
-  /// Generates an import-only stylesheet for [entrypoint] that forwards any
-  /// members that used to have a prefix with that prefix, but forwards other
-  /// members as-is.
+  /// Generates an import-only file for [entrypoint].
   ///
-  /// If there are no previously-prefixed members to forward, this returns null.
-  String _generateImportOnly(Uri entrypoint, String dependency) {
-    var semicolon = entrypoint.path.endsWith('.sass') ? '' : ';';
-    var forwardWithPrefix = <MemberDeclaration>{};
-    var forwardWithoutPrefix = <MemberDeclaration>{};
+  /// If a prefix was removed from any members, this will add that prefix back
+  /// for the import-only file. If `--forward=import-only` was passed, this will
+  /// also add `@forward` rules for members in dependencies that aren't already
+  /// forwarded through the entrypoint itself.
+  ///
+  /// If there are no previously-prefixed members or members from dependencies
+  /// to forward, this returns null.
+  String _generateImportOnly(Uri entrypoint, Uri importOnlyUrl) {
+    // Sort all members based on the URL they should be forwarded from and the
+    // prefix they require (if any).
+    var forwardsByUrl = <Uri, Map<String, Set<MemberDeclaration>>>{};
+    var hiddenByUrl = <Uri, Set<MemberDeclaration>>{};
     for (var declaration in references.globalDeclarations) {
+      var private = declaration.name.startsWith('-');
       var visibleAtEntrypoint = declaration.sourceUrl == entrypoint ||
-          (_shouldForward(declaration.name) &&
-              !declaration.name.startsWith('-'));
-      if (!visibleAtEntrypoint) continue;
-
-      if (declaration.name.startsWith(prefixToRemove)) {
-        forwardWithPrefix.add(declaration);
-      } else {
-        forwardWithoutPrefix.add(declaration);
+          (_shouldForward(declaration.name) && !private);
```

Comment:
It's not immediately clear to me what the meaning/use of this variable is. Can you add a comment?